scheme/nvidia-coserv/coserv_handler.go

```diff
@@ -0,0 +1,225 @@
+
+	if resp.StatusCode != http.StatusOK {
+		body, _ := io.ReadAll(resp.Body)
+		return nil, fmt.Errorf("error response from NVIDIA RIM service: %d %s\nResponse body: %s", resp.StatusCode, resp.Status, string(body))
+	}
+
+	// Read the successful response
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("reading response error: %w", err)
+	}
+
+	// Successful responses are JSON, modelled by the RimServiceResponse class,
+	// so deserialize into that.
+	var rimServiceResponse RimServiceResponse
+	err = json.Unmarshal(body, &rimServiceResponse)
+	if err != nil {
+		return nil, fmt.Errorf("unmarshalling response error: %w", err)
+	}
+
+	return &rimServiceResponse, nil
+}
+
+func (s CoservProxyHandler) addReferenceValuesForClass(query *coserv.Query, c *coserv.StatefulClass, results *coserv.ResultSet) error {
+	if *c.Class.Vendor != "NVIDIA" {
+		return fmt.Errorf("vendors other than NVIDIA not supported by this proxy plug-in")
+	}
+
+	if c.Measurements != nil {
+		return fmt.Errorf("NVIDIA CoSERV proxy plug-in does not expect stateful class environments")
+	}
+
+	if c.Class.Model == nil {
+		return fmt.Errorf("no Model field supplied in NVIDIA class environment")
+	}
+
+	rimid := c.Class.Model
+
+	rimServiceResponse, err := callRimService(rimid)
+
+	if err != nil {
+		return err
+	}
+
+	// The actual RIM is a base64-encoded field, so pull this out as a byte array.
+	rimBytes, err := base64.StdEncoding.DecodeString(rimServiceResponse.Rim)
+
+	if err != nil {
+		return fmt.Errorf("failed to base64 decode the RIM byte string %s", rimServiceResponse.Rim)
+	}
+
+	// If we want the raw source artifacts, add the RIM verbatim into the result set.
+	// We just need to switch on the RIM type (TCG or CORIM), and use a different media type
+	// label in each case.
+	if query.ResultType == coserv.ResultTypeBoth || query.ResultType == coserv.ResultTypeSourceArtifacts {
+		switch rimServiceResponse.RimFormat {
+		case "CORIM":
+			cmw, err := cmw.NewMonad("application/rim+cbor", rimBytes)
+			if err != nil {
+				return fmt.Errorf("failed to create CMW for CORIM source artifact: %w", err)
+			}
+			results.AddSourceArtifacts(*cmw)
+		case "TCG":
+			cmw, err := cmw.NewMonad("application/swid+xml", rimBytes)
+			if err != nil {
+				return fmt.Errorf("failed to create CMW for TCG source artifact: %w", err)
+			}
+			results.AddSourceArtifacts(*cmw)
+		default:
+			return fmt.Errorf("unexpected RIM format %s from NVIDIA RIM service", rimServiceResponse.RimFormat)
+		}
+	}
+
+	// If we want collected artifacts, we need to parse the CORIM
+	if query.ResultType == coserv.ResultTypeBoth || query.ResultType == coserv.ResultTypeCollectedArtifacts {
+		if rimServiceResponse.RimFormat != "CORIM" {
+			return fmt.Errorf("CoSERV proxy plug-in cannot produce collected artifacts for non-CORIM format %s", rimServiceResponse.RimFormat)
+		}
+
+		// We now know that the format is CORIM, so it should be possible to decode as such
```

Comment:
An architectural point, worth discussing via a github issue, may be I will track it in the repo..

If NVIDIA is already reporting a CoRIM which is signed, I see no point breaking it down, re-formatting it.

Also potentially I would be worried, even more, if we lose some information from CoRIM in the response!

Yes, there is a facility to stash the entire blob as a side object to the re-package, if I understand CoSERV correctly, then however, this sounds like an over-kill !

Reply:
There is no architectural issue here; this is exactly how CoSERV is supposed to work.  It's more of a RATS discussion around the CoSERV spec that should be done over the RATS ML.